src/dbsync/utils/threadDispatcher.h

```diff
@@ -0,0 +1,164 @@
+	 
+	// template <typename Type, typename Functor>
+	// class DispatcherInterface
+	// {
+	// public:
+	// 	/**
+	// 	 * @brief Ctor
+	// 	 * 
+	// 	 * @param functor Callable entity.
+	// 	 * @param int Maximun number of threads to be used by the dispatcher.
+	// 	 */
+	// 	DispatcherInterface(Functor functor, const unsigned int numberOfThreads);
+	// 	*
+	// 	 * @brief Pushes a message to be processed by the functor.
+	// 	 * @details The implementation decides whether the processing is sync or async.
+	// 	 * 
+	// 	 * @param data Message value.
+		 
+	// 	void push(const Type& data);
+	// 	/**
+	// 	 * @brief Rundowns the pending messages until reaches 0.
+	// 	 * @details It should be a blocking call.
+	// 	 */		
+	// 	void rundown();
+	// 	/**
+	// 	 * @brief Cancels the dispatching.
+	// 	 */
+	// 	void cancel();
+	// };
+
+	template
+	<
+	typename Type,
+	typename Functor
+	>
+	class AsyncDispatcher
+	{
+	public:
+		AsyncDispatcher(Functor functor, const unsigned int numberOfThreads = std::thread::hardware_concurrency())
+		: m_functor{ functor }
+		, m_blocked{ false }
+		, m_numberOfThreads{ numberOfThreads }
+		{
+			m_threads.reserve(m_numberOfThreads);
+			for (unsigned int i = 0; i < m_numberOfThreads; ++i)
+			{
+				m_threads.push_back(std::thread{ &AsyncDispatcher<Type, Functor>::dispatch, this });
+			}
+		}
+		AsyncDispatcher& operator=(const AsyncDispatcher&) = delete;
+		AsyncDispatcher(AsyncDispatcher& other) = delete;
+		~AsyncDispatcher()
+		{
+			m_queue.cancel();
+			for (auto& thread : m_threads)
+			{
+				thread.join();
+			}
+		}
+
+		void push(const Type& value)
+		{
+			if (!m_blocked)
+			{
+				m_queue.push
+				(
+					[value, this]()
+					{
+						this->m_functor(value);
+					}
+				);
+			}
+		}
+
+		void rundown()
+		{
+			if (!cancelled())
+			{
+				m_blocked = true;
+				std::promise<void> promise;
+				auto fut { promise.get_future() };
+				m_queue.push
+				(
+					[&promise]()
+					{
+						promise.set_value();
+					}
+				);
+				fut.wait();
+				cancel();
+			}
+		}
+		void cancel()
+		{
+			m_queue.cancel();
+		}
+
+		bool cancelled() const
+		{
+			return m_queue.cancelled();
+		}
+		unsigned int numberOfThreads() const
+		{
+			return m_numberOfThreads;
+		}
+	    size_t size() const
+	    {
+	    	return m_queue.size();
+	    }
+
+	private:
+		void dispatch()
+		{
+			std::function<void()> fnc;
+			while(m_queue.pop(fnc))
+			{
+				fnc();
+			}
+		}
+		Functor m_functor;
+		SafeQueue<std::function<void()>> m_queue;
+		std::vector<std::thread> m_threads;
+		std::atomic_bool m_blocked;
+		const unsigned int m_numberOfThreads;
+	};
+
+	template <typename Input, typename Functor>
+	class SyncDispatcher
+	{
+	public:
+	    SyncDispatcher(Functor functor, const unsigned int numberOfThreads = 0)
+	    : m_functor{functor}
+	    {
+	    }
+	    void push(const Input& data)
+	    {
+	        m_functor(data);
+	    }
+	    size_t size() const {return 0;}
```

Comment:
Questions:

Are these methods necessary? I don't see inheritance from a pure abstract class.

Could you make a base class called "Dispatcher" for ASyncDispatcher and SyncDispatcher?

Reply:
using a base class as an interface will force the use of dynamic polymorphism and the intend here is to use policy classes and static polymorphism. Any class that exposes the methods described in lines 23-57 can be used as a thread dispatching policy.